util/gpool/spmc/spmspool.go

```diff
@@ -0,0 +1,385 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spmc
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/log"
+	"github.com/pingcap/tidb/resourcemanager/pooltask"
+	"github.com/pingcap/tidb/util/gpool"
+	"go.uber.org/zap"
+)
+
+// Pool is a single producer, multiple consumer goroutine pool.
+type Pool[T any, U any, C any, CT any, TF pooltask.Context[CT]] struct {
+	gpool.BasePool
+	workerCache   sync.Pool
+	workers       *loopQueue[T, U, C, CT, TF]
+	lock          sync.Locker
+	cond          *sync.Cond
+	taskCh        chan *pooltask.TaskBox[T, U, C, CT, TF]
+	options       *Options
+	stopCh        chan struct{}
+	consumerFunc  func(T, C, CT) U
+	capacity      atomic.Int32
+	running       atomic.Int32
+	state         atomic.Int32
```

Comment:
Could we use a field to handle these 3 states(`OPENED`, `CLOSED`, and `RUNNING`)?  Or using the `bool` type instead of `int32` for `running`